doc/man/man5/sge_ckpt.5
'\" t
.\"___INFO__MARK_BEGIN__
.\"
.\" Copyright: 2004 by Sun Microsystems, Inc.
.\"
.\"___INFO__MARK_END__
.\"
.\"
.\" Some handy macro definitions [from Tom Christensen's man(1) manual page].
.\"
.de SB		\" small and bold
.if !"\\$1"" \\s-2\\fB\&\\$1\\s0\\fR\\$2 \\$3 \\$4 \\$5
..
.\"
.de T		\" switch to typewriter font
.ft CW		\" probably want CW if you don't have TA font
..
.\" "
.de TY		\" put $1 in typewriter font
.if t .T
.if n ``\c
\\$1\c
.if t .ft P
.if n \&''\c
\\$2
..
.\"
.de M		\" man page reference
\\fI\\$1\\fR\\|(\\$2)\\$3
..
.TH xxQS_NAME_Sxx_CKPT 5 2012-09-18 "xxRELxx" "xxQS_NAMExx User Commands"
.\"
.SH NAME
xxqs_name_sxx_ckpt \- the xxQS_NAMExx checkpointing mechanism and checkpointing
support
.\"
.SH DESCRIPTION
xxQS_NAMExx
supports two levels of checkpointing: the user level and an operating
system-provided transparent
level. User level checkpointing refers to applications which do their
own checkpointing by writing restart files at certain times or
algorithmic steps and by properly processing these restart files when
restarted.
.PP
Transparent checkpointing has to be provided by the operating system and is 
usually integrated in the operating system kernel. An example for a kernel 
integrated checkpointing facility is the Hibernator package from Softway
for SGI IRIX platforms.
.PP
Checkpointing jobs need to be identified to the xxQS_NAMExx system by using the 
\fI\-ckpt\fP option of the
.M qsub 1
command. The argument to this flag refers to a so 
called checkpointing environment, which defines the attributes of the 
checkpointing method to be used (see
.M checkpoint 5
for details). 
Checkpointing environments are setup by the
.M qconf 1
options \fI\-ackpt\fP, \fI\-dckpt\fP, \fI\-mckpt\fP and \fI\-sckpt\fP. The
.M qsub 1
option \fI\-c\fP can be used to overwrite the \fIwhen\fP
attribute for the referenced checkpointing environment.
.PP
As opposed to the behavior for
regular batch jobs, checkpointing jobs (see the \fB\-ckpt\fP option to
.M qsub 1 )
are aborted under conditions
for which batch or interactive jobs are suspended or even stay
unaffected. These conditions are:
.\"
.IP "\(bu" 3n
Explicit suspension of the queue or job via
.M qmod 1
by the cluster administration or a queue owner
if the \fIx\fP occasion specifier (see
.M qsub 1
\fI\-c\fP and 
.M checkpoint 5 )
was assigned to the job.
.\"
.IP "\(bu" 3n
A load average value exceeding the suspend threshold as configured for
the corresponding queues (see
.M queue_conf 5 ).
.\"
.IP "\(bu" 3n
Shutdown of the xxQS_NAMExx execution daemon
.M xxqs_name_sxx_execd 8
being responsible for the checkpointing job.
.PP
.\"
After they are aborted, jobs will migrate to other hosts, and possibly
other cluster queues, unless they were
submitted to a specific one by an explicit user request.
The migration of jobs leads to a dynamic load balancing.
\fBNote:\fP Aborting checkpointed jobs will free all resources
(memory, swap space) which the job occupies at that time. This is
opposed to the situation for suspended regular jobs, which still use
virtual memory and other consumable resources.
.PP
.\"
.\"
.SH RESTRICTIONS
When a job migrates to another machine, at present no files
are transferred automatically to that machine. This means that all files
which are used throughout the entire job, including restart files,
executables, and scratch files, must be visible or transferred explicitly
(e.g. at the beginning of the job script).
.PP
.\"
There are also some practical limitations regarding use of disk space
for transparently checkpointing jobs. Checkpoints of a transparently
checkpointed application are usually stored in a checkpoint file or
directory by the operating system. The file or directory contains all
the text, data, and stack space for the process, along with some
additional control information. This means jobs which use a very large
virtual address space will generate very large checkpoint files. Also
the workstations on which the jobs will actually execute may have
little free disk space. Thus it is not always possible to transfer a
transparent checkpointing job to a machine, even though that machine is
idle. Since large virtual memory jobs must wait for a machine that is
both idle, and has a sufficient amount of free disk space, such jobs
may suffer long turnaround times.
.PP
There is currently no mechanism for restarting jobs with the same resources they
were granted originally.  That might be important if they were
submitted with a choice or range of resources and start running in a
particular way with what they're given.
.PP
Similarly, with heterogeneous execution hosts, jobs may need to
restart on a host which supports a superset of the instruction set
where the job originally ran if the checkpoint mechanism (e.g. BLCR or
DMTCP) dumps an image of the running process.  Runtime libraries, in
particular, may initialize themselves depending on details of the
architecture they start up on \- say to use a specific type of vector
unit.  Then, they may fail if moved to an older host of similar
architecture which lacks that feature, even if they were compiled for
a common instruction set.
.\"
.SH "SEE ALSO"
.M xxqs_name_sxx_intro 1 ,
.M qconf 1 ,
.M qmod 1 ,
.M qsub 1 ,
.M checkpoint 5
.\" .M checkpoint 5 ,
.\" .I xxQS_NAMExx Installation and Administration Guide,
.\" .I xxQS_NAMExx User's Guide
.\"
.SH "COPYRIGHT"
See
.M xxqs_name_sxx_intro 1
for a full statement of rights and permissions.


NEWS
Notable changes in Son of Grid Engine
=====================================

For earlier changes (comprising Sun Grid Engine), see Changelog in the
source distribution, and the bug lists at
http://arc.liv.ac.uk/SGE/howto/Installation,%20Upgrade,%20Patches.
For detailed recent change information, including credits for changes
not listed here, see https://arc.liv.ac.uk/trac/SGE/log/sge/

"[#<n>]" below refers to ticket <n> at
https://arc.liv.ac.uk/trac/SGE/ticket/<n>.  For historical reasons,
changes due to Univa/Sun are tagged as "[U]", or "[(U)]" if partially
due to them or appearing first in this version.


Version 8.1.9
-------------

* Security
  * Drop setgid privileges correctly (CERT POS36-C); daemons now run
    with (only) admin user's groups, not root's

* Building

  * MUNGE authentication can be built in (aimk -munge)
  * Use a minimal autoconf framework
  * Fix building on recent Darwin and Cygwin [#1516, #1536, #1557]
  * Drop support for Darwin < 10 and Darwin/PPC
  * Cope with GNU/Linux ppc64le as well as ppc64el
  * Fix rpm packaging of spool objects [#1519]
  * Actually install "work" in utilbin (c.f. 8.1.8)
  * Rpm spec fixes for Fedora 23, ppc64, and OpenSuSE
  * Fix build when bool is defined in headers

* Bug fixes

  * Man fixes
  * Message fixes [#1538, #1539]
  * Fix RPM packaging of spool files [#1519]
  * Don't set port and cluster in check_sge.py, and accept memory in T
  * Fix pam_sge-qrsh-setup for recent PAM
  * Do a full, not a partial match on GECOS data when selecting certificate
    in sge_ca [#1517]
  * Fix checking of admin host in qconf [#1547]
  * Report the actual exit status of master task in tightly integrated
    parallel jobs
  * Fix share tree accounting with array jobs [#435]
  * Fix doubled multipliers of units in qacct memory reports
  * Improve job scheduling within a sharetree node [#1550]
  * Flush spool at qmaster exit [#1551]
  * Sort jobs and advance reservations when reading from spool [#1552, #1553]
  * Fix detection of old v. new type of cpuset [#1523]
  * Fix treatment of some invalid entries in variable lists or the
    environment on job submission

* Enhancements

  * MUNGE authentication (as above)
  * qconf -clearusage can act on a specific user
  * Allow generating load sensor data just before sending them
  * Add sge_ca "-revoke <cert>" option
  * Allow running DRMAA jobs as arbitrary user [#492].  (Partially
    done in v 8.1.0, but not mentioned; now works with CSP.)
  * Report SSL error numbers in hex [#1539]
  * Use PVM_VMID environment variable [#44]
  * rc templates support a "status" argument (for HA usage)

* Potentially incompatible changes

  * Print error messages to stderr and don't print on success with -terse
    to be consistent with submit(1) saying that error messages go to
    stderr and to make -terse more useful in cases like
      cat test.o$(qsub -sync y -terse -cwd -N test ...)
  * sge_ca -renew and friends no longer revoke old certificates
  * The qconf -sep and -cb options have been removed finally
  * A POSIX 2001-compatible library is required

Version 8.1.8
-------------

* Building

  * Fix build with recent GNU binutils
  * Fix rpm build on Fedora and maybe use packaged swing-layout
  * Fix dpkg build on recent Debian/Ubuntu
  * uid_t/gid_t are unsigned on NetBSD
  * Install "work" in utilbin
  * Move qsched back to main package

* Bug fixes

  * Man fixes
  * Fix buffer overflow in qmake [#1507]
  * Avoid possible link-time error (Debian bug # 749413)
  * Fix some global messages from qstat -j
  * Avoid truncating usage values in accounting records
  * Fix execd remembering existing core binding when restarted [#1511]

* Enhancements

  * Nagios and Ganglia monitoring scripts
  * Python JSV implementation
  * Upstart and systemd startup files
  * Define DRMAA_LIBRARY_PATH in setup scripts

Version 8.1.7
-------------

* Building

  * Remove the shipped old jemalloc in favour of an external one
    (aimk -jemalloc)
  * Always use system libraries -- never load SSL stuff dynamically
  * aimk no longer uses specific paths for openssl etc.  Use
    INPUT_*_FLAGS if necessary
  * SUpport building rpm on OpenSuSE and RHEL7

* Bug fixes

  * Man, other doc, and message fixes/changes [#1504]
  * Fix minor memory leaks and check result of all malloc calls
  * Don't pause for a second with client JSVs [#1475]
  * Allow longer "failed" strings in admin mail
  * Fix spurious thread error reports from qping [#588]
  * Avoid possible crash reading malformed configuration files
  * Fix GUI installer startup [#1488]
  * Always define SGE_STARTER_USE_LOGIN_SHELL with starter method
  * Modify cpuset-based process killing [#1483, #1490]
  * Make various files optional when setting permissions at installation
  * Fix cpu-bound shepherd with qlogin
  * Fix queue list for adding queue in qmon [#1495]
  * Clean up some file permissions and shebangs in non-executables [#1493]
  * Really ignore -e, -o args of qrsh/qsh/qlogin, as documented
  * Fix use of SGE_QMASTER_PIDFILE and double writing of pid file
  * Fix occasional qmaster major space leak with schedd_job_info [#360, #682]
  * Wait until daemon goes down when stopping sge_execd

* Enhancements

  * Support Cygwin x86_64, and GNU/Linux on arm64 and ppc64el (untested)
  * qconf -aattr et al now support modifying types user, userset, calendar,
    and project
  * Translate exit code to signal name in qacct output
  * qacct -m prints only master tasks
  * Add ar_sub_time to qacct output
  * Allow rctemplate configuration from /etc/sysconfig as well as
    /etc/default
  * Update MPI PE templates and install a tight integration "mpi" PE
  * Make suid files, dtrace, qmon, start_gui_installer optional in
    setfileperm.sh (used at installation)
  * Speed up JSV in bash by not calling cut

* Potentially incompatible changes

  * Default ENABLE_ADDGRP_KILL to on
  * Clean symbol exports from libdrmaa.so (GNU ld only); now only exports

source/libs/sgeobj/sge_ckpt.c
/*___INFO__MARK_BEGIN__*/
/*************************************************************************
 * 
 *  The Contents of this file are made available subject to the terms of
 *  the Sun Industry Standards Source License Version 1.2
 * 
 *  Sun Microsystems Inc., March, 2001
 * 
 * 
 *  Sun Industry Standards Source License Version 1.2
 *  =================================================
 *  The contents of this file are subject to the Sun Industry Standards
 *  Source License Version 1.2 (the "License"); You may not use this file
 *  except in compliance with the License. You may obtain a copy of the
 *  License at http://gridengine.sunsource.net/Gridengine_SISSL_license.html
 * 
 *  Software provided under this License is provided on an "AS IS" basis,
 *  WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING,
 *  WITHOUT LIMITATION, WARRANTIES THAT THE SOFTWARE IS FREE OF DEFECTS,
 *  MERCHANTABLE, FIT FOR A PARTICULAR PURPOSE, OR NON-INFRINGING.
 *  See the License for the specific provisions governing your rights and
 *  obligations concerning the Software.
 * 
 *   The Initial Developer of the Original Code is: Sun Microsystems, Inc.
 * 
 *   Copyright: 2001 by Sun Microsystems, Inc.
 * 
 *   All Rights Reserved.
 * 
 ************************************************************************/
/*___INFO__MARK_END__*/

#include <strings.h>
#include <ctype.h>

#include "uti/sge_rmon.h"
#include "uti/sge_log.h"
#include "uti/config_file.h"
#include "uti/sge_signal.h"

#include "cull/cull_list.h"

#include "sgeobj/sge_object.h"
#include "sgeobj/sge_answer.h"
#include "sgeobj/sge_job.h"
#include "sgeobj/sge_cqueue.h"
#include "sgeobj/sge_attr.h"
#include "sgeobj/sge_qinstance.h"
#include "sgeobj/sge_utility.h"
#include "sgeobj/sge_ckpt.h"
#include "sgeobj/sge_str.h"
#include "sgeobj/msg_sgeobjlib.h"

#include "symbols.h"
#include "msg_common.h"

/****** sgeobj/ckpt/ckpt_is_referenced() **************************************
*  NAME
*     ckpt_is_referenced() -- Is a given CKPT referenced in other objects? 
*
*  SYNOPSIS
*     bool ckpt_is_referenced(const lListElem *ckpt, lList **answer_list, 
*                             const lList *master_job_list,
*                             const lList *master_cqueue_list) 
*
*  FUNCTION
*     This function returns true if the given "ckpt" is referenced
*     in at least one of the objects contained in "master_job_list" or
*     "master_cqueue_list". If this is the case than
*     a corresponding message will be added to the "answer_list". 
*
*  INPUTS
*     const lListElem *ckpt           - CK_Type object 
*     lList **answer_list             - AN_Type list 
*     const lList *master_job_list    - JB_Type list 
*     const lList *master_cqueue_list - CQ_Type list
*
*  RESULT
*     bool - true or false  
******************************************************************************/
bool ckpt_is_referenced(const lListElem *ckpt, lList **answer_list,
                        const lList *master_job_list, 
                        const lList *master_cqueue_list)
{
   bool ret = false;

   {
      lListElem *job = NULL;

      for_each(job, master_job_list) {
         if (job_is_ckpt_referenced(job, ckpt)) {
            const char *ckpt_name = lGetString(ckpt, CK_name);
            u_long32 job_id = lGetUlong(job, JB_job_number);

            answer_list_add_sprintf(answer_list, STATUS_EUNKNOWN,
                                    ANSWER_QUALITY_INFO, MSG_CKPTREFINJOB_SU,
                                    ckpt_name, sge_u32c(job_id));
            ret = true;
            break;
         }
      } 
   }
   if (!ret) {
      lListElem *queue = NULL, *ckl = NULL;
 
      /* fix for bug 6422335
       * check the cq configuration for ckpt references instead of qinstances
       */
      const char *ckpt_name = lGetString(ckpt, CK_name);

      for_each(queue, master_cqueue_list) {
         for_each(ckl, lGetList(queue, CQ_ckpt_list)){
            if (lGetSubStr(ckl, ST_name, ckpt_name, ASTRLIST_value))  {
               answer_list_add_sprintf(answer_list, STATUS_EUNKNOWN,
                                       ANSWER_QUALITY_INFO, 
                                       MSG_CKPTREFINQUEUE_SS,
                                       ckpt_name, lGetString(queue, CQ_name));
               ret = true;
               break;
            }
         }
      }
   }
   return ret;
}

/****** sgeobj/ckpt/ckpt_list_locate() ***************************************
*  NAME
*     ckpt_list_locate -- find a ckpt object in a list 
*
*  SYNOPSIS
*     lListElem *ckpt_list_locate(lList *ckpt_list, const char *ckpt_name)
*
*  FUNCTION
*     This function will return a ckpt object by name if it exists.
*
*
*  INPUTS
*     lList *ckpt_list      - CK_Type object
*     const char *ckpt_name - name of the ckpt object. 
*
*  RESULT
*     NULL - ckpt object with name "ckpt_name" does not exist
*     !NULL - pointer to the cull element (CK_Type) 
******************************************************************************/
lListElem *ckpt_list_locate(const lList *ckpt_list, const char *ckpt_name)
{
   return lGetElemStr(ckpt_list, CK_name, ckpt_name);
}

/****** sgeobj/ckpt/sge_parse_checkpoint_attr() *******************************
*  NAME
*     sge_parse_checkpoint_attr() -- make "when" bitmask from string 
*
*  SYNOPSIS
*     int sge_parse_checkpoint_attr(const char *attr_str) 
*
*  FUNCTION
*     Parse checkpoint "when" string and return a bitmask. 
*
*  INPUTS
*     const char *attr_str - when string 
*
*  RESULT
*     int - bitmask of checkpoint specifers
*           0 if attr_str == NULL or nothing set or value 
*           may be a time value 
*
*  NOTES
*     MT-NOTE: sge_parse_checkpoint_attr() is MT safe
*******************************************************************************/
int sge_parse_checkpoint_attr(const char *attr_str)
{
   int opr;

   if (attr_str == NULL) {
      return 0;
   }

   /* May be it's a time value */
   if (isdigit((int) *attr_str) || (*attr_str == ':')) {
      return 0;
   }

   opr = 0;
   while (*attr_str) {
      if (*attr_str == CHECKPOINT_AT_MINIMUM_INTERVAL_SYM)
         opr = opr | CHECKPOINT_AT_MINIMUM_INTERVAL;
      else if (*attr_str == CHECKPOINT_AT_SHUTDOWN_SYM)
         opr = opr | CHECKPOINT_AT_SHUTDOWN;
      else if (*attr_str == CHECKPOINT_SUSPEND_SYM)
         opr = opr | CHECKPOINT_SUSPEND;
      else if (*attr_str == NO_CHECKPOINT_SYM)
         opr = opr | NO_CHECKPOINT;
      else if (*attr_str == CHECKPOINT_AT_AUTO_RES_SYM)
         opr = opr | CHECKPOINT_AT_AUTO_RES;
      else {
         opr = -1;
         break;
      }
      attr_str++;
   }

   return opr;
}

/****** sgeobj/ckpt/ckpt_validate() ******************************************
*  NAME
*     ckpt_validate -- validate all ckpt interface parameters
*
*  SYNOPSIS
*     int ckpt_validate(lListElem *ep, lList **alpp);
*
*  FUNCTION
*     This function will test all ckpt interface parameters.
*     If all are valid then it will return successfull.
*
*  INPUTS
*     ep     - element which sould be verified.
*     answer - answer list where the function stored error messages
*
*  RESULT
*     [answer] - error messages will be added to this list
*     STATUS_OK - success
*     STATUS_EUNKNOWN or STATUS_EEXIST - error
*
*  NOTES
*     MT-NOTE: ckpt_validate() is not MT safe
******************************************************************************/
int ckpt_validate(const lListElem *this_elem, lList **alpp)
{
   static const char* ckpt_interfaces[] = {
      "USERDEFINED",
      "HIBERNATOR",
      "TRANSPARENT",
      "APPLICATION-LEVEL",
      "CPR"
   };
   static struct attr {
      int nm;
      char *text;
   } ckpt_commands[] = {
      { CK_ckpt_command, "ckpt_command" },
      { CK_migr_command, "migr_command" },
      { CK_rest_command, "restart_command"},
      { CK_clean_command, "clean_command"},
      { NoName,           NULL} };

   unsigned i;
   int found = 0;
   const char *s, *interface;

   DENTER(TOP_LAYER, "ckpt_validate");

   if (!this_elem) {
      CRITICAL((SGE_EVENT, MSG_SGETEXT_NULLPTRPASSED_S, SGE_FUNC));
      answer_list_add(alpp, SGE_EVENT, STATUS_EUNKNOWN, ANSWER_QUALITY_ERROR);
      DEXIT;
      return STATUS_EUNKNOWN;
   }

   /* -------- CK_name */
   if (verify_str_key(
         alpp, lGetString(this_elem, CK_name), 
         MAX_VERIFY_STRING, "checkpoint interface", KEY_TABLE) != STATUS_OK) {
      DEXIT;
      return STATUS_EUNKNOWN;
   }

   /*
    * check if ckpt obj can be added
    * check allowed interfaces and license
    */
   interface = lGetString(this_elem, CK_interface);
   found = 0;
   /* handle NULL pointer for interface name */
   if (interface == NULL) {
      interface = "<null>";
   } else {

      for (i=0; i < (sizeof(ckpt_interfaces)/sizeof(char*)); i++) {
         if (!strcasecmp(interface, ckpt_interfaces[i])) {
            found = 1;
            break;
         }
      }
   }

   if (!found) {
      ERROR((SGE_EVENT, MSG_SGETEXT_NO_INTERFACE_S, interface));
      answer_list_add(alpp, SGE_EVENT, 
                      STATUS_ESEMANTIC, ANSWER_QUALITY_ERROR);
      DEXIT;
      return STATUS_EEXIST;
   }

   for (i = 0; ckpt_commands[i].nm != NoName; i++) {
      if (replace_params(lGetString(this_elem, ckpt_commands[i].nm),
               NULL, 0, ckpt_variables)) {
         ERROR((SGE_EVENT, MSG_OBJ_CKPTENV_SSS,
               ckpt_commands[i].text, lGetString(this_elem, CK_name), err_msg));
         answer_list_add(alpp, SGE_EVENT, STATUS_EEXIST, ANSWER_QUALITY_ERROR);
         DEXIT;
         return STATUS_EEXIST;
      }
   }

   /* -------- CK_signal */
   if ((s=lGetString(this_elem, CK_signal)) &&
         strcasecmp(s, "none") &&
         sge_sys_str2signal(s)==-1) {
      ERROR((SGE_EVENT, MSG_CKPT_XISNOTASIGNALSTRING_S , s));
      answer_list_add(alpp, SGE_EVENT, STATUS_EEXIST, ANSWER_QUALITY_ERROR);
      DEXIT;
      return STATUS_EEXIST;
   }

   DEXIT;
   return STATUS_OK;
}

/****** sgeobj/ckpt/ckpt_list_get_master_list() *******************************
*  NAME
*     ckpt_list_get_master_list() -- Return pointer to master ckpt list 
*
*  SYNOPSIS
*     lList ** ckpt_list_get_master_list(void) 
*
*  FUNCTION
*     Return pointer to master ckpt list 
*
*  RESULT
*     lList ** - master ckpt list
*******************************************************************************/
lList **
ckpt_list_get_master_list(void)
{
   return object_type_get_master_list(SGE_TYPE_CKPT);
}

/****** sgeobj/ckpt/ckpt_list_do_all_exist() **********************************
*  NAME
*     ckpt_list_do_all_exist() -- Do all ckpt's exist? 
*
*  SYNOPSIS
*     bool 
*     ckpt_list_do_all_exist(const lList *ckpt_list, 
*                            lList **answer_list, 
*                            const lList *ckpt_ref_list) 
*
*  FUNCTION
*     Test if the checkpointing objects whose name is contained in
*     "ckpt_ref_list" is contained in "ckpt_list". 
*
*  INPUTS
*     const lList *ckpt_list     - CK_Type list 
*     lList **answer_list        - AN_Type list 
*     const lList *ckpt_ref_list - ST_Type list containing ckpt names 
*
*  RESULT
*     bool - true if all ckpt objects exist 
*******************************************************************************/
bool
ckpt_list_do_all_exist(const lList *ckpt_list, lList **answer_list,
                       const lList *ckpt_ref_list)
{
   bool ret = true;
   lListElem *ckpt_ref_elem = NULL;

   DENTER(TOP_LAYER, "ckpt_list_do_all_exist");
   for_each(ckpt_ref_elem, ckpt_ref_list) {
      const char *ckpt_ref_string = lGetString(ckpt_ref_elem, ST_name);

      if (ckpt_list_locate(ckpt_list, ckpt_ref_string) == NULL) {
         answer_list_add_sprintf(answer_list, STATUS_EEXIST,
                                 ANSWER_QUALITY_ERROR,
                                 MSG_CKPTREFDOESNOTEXIST_S, ckpt_ref_string);
         ret = false;
         break;
      }
   }
   DEXIT;
   return ret;
}


/****** src/sge_generic_ckpt() **********************************************
*
*  NAME
*     sge_generic_ckpt -- build up a generic ckpt object
*
*  SYNOPSIS
*     lListElem* sge_generic_ckpt(
*        char *ckpt_name
*     );
*
*  FUNCTION
*     build up a generic ckpt object
*
*  INPUTS
*     ckpt_name - name used for the CK_name attribute of the generic
*               pe object. If NULL then "template" is the default name.
*
*  RESULT
*     !NULL - Pointer to a new CULL object of type CK_Type
*     NULL - Error
*
*  EXAMPLE
*
*  NOTES
*
*  BUGS
*
*  SEE ALSO
*
*****************************************************************************/   
lListElem* sge_generic_ckpt(char *ckpt_name) 
{
   lListElem *ep;

   DENTER(TOP_LAYER, "sge_generic_ckpt");

   ep = lCreateElem(CK_Type);

   if (ckpt_name)
      lSetString(ep, CK_name, ckpt_name);
   else
      lSetString(ep, CK_name, "template");

   lSetString(ep, CK_interface, "userdefined");
   lSetString(ep, CK_ckpt_command, "none");
   lSetString(ep, CK_migr_command, "none");
   lSetString(ep, CK_rest_command, "none");
   lSetString(ep, CK_clean_command, "none");
   lSetString(ep, CK_ckpt_dir, "/tmp");
   lSetString(ep, CK_when, "sx");
   lSetString(ep, CK_signal, "none");
   lSetUlong(ep, CK_job_pid, 0);

   DEXIT;
   return ep;
}

